htscodecs/rANS_static4x16pr.c

```diff
@@ -1114,24 +746,329 @@ unsigned char *rans_uncompress_O1_4x16(unsigned char *in, unsigned int in_size,
+
+	int level = __get_cpuid_max(0, NULL);
+	if (level >= 1) {
+	    __cpuid_count(1, 0, eax, ebx, ecx, edx);
+#if defined(bit_SSSE3)
+	    have_ssse3 = ecx & bit_SSSE3;
+#endif
+#if defined(bit_POPCNT)
+	    have_popcnt = ecx & bit_POPCNT;
+#endif
+#if defined(bit_SSE4_1)
+	    have_sse4_1 = ecx & bit_SSE4_1;
+#endif
+	}
+	if (level >= 7) {
+	    __cpuid_count(7, 0, eax, ebx, ecx, edx);
+#if defined(bit_AVX2)
+	    have_avx2 = ebx & bit_AVX2;
+#endif
+#if defined(bit_AVX512F)
+	    have_avx512f = ebx & bit_AVX512F;
+#endif
+	}
+
+	if (!have_popcnt) have_avx512f = have_avx2 = have_sse4_1 = 0;
+	if (!have_ssse3)  have_sse4_1 = 0;
+
+	if (!(rans_cpu & RANS_CPU_DEC_AVX512)) have_avx512f = 0;
+	if (!(rans_cpu & RANS_CPU_DEC_AVX2))   have_avx2 = 0;
+	if (!(rans_cpu & RANS_CPU_DEC_SSE4))   have_sse4_1 = 0;
+
+//	fprintf(stderr, "SSSE3 %d, SSE4.1 %d, POPCNT %d, AVX2 %d, AVX512F %d\n",
+//		have_ssse3, have_sse4_1, have_popcnt, have_avx2, have_avx512f);
+
+	if (order & 1) {
+	    return have_avx512f
+		? rans_uncompress_O1_32x16_avx512
+		: (have_avx2
+		   ? rans_uncompress_O1_32x16_avx2
+		   : (have_sse4_1
+		      ? rans_uncompress_O1_32x16_sse4
+		      : rans_uncompress_O1_32x16));
+	} else {
+	    return have_avx512f
+		? rans_uncompress_O0_32x16_avx512
+		: (have_avx2
+		   ? rans_uncompress_O0_32x16_avx2
+		   : (have_sse4_1
+		      ? rans_uncompress_O0_32x16_sse4
+		      : rans_uncompress_O0_32x16));
+	}
+    }
+
+    // Else all SIMD disabled
+    return order & 1
+	? rans_uncompress_O1_4x16
+	: rans_uncompress_O0_4x16;
+}
+
+#elif defined(__ARM_NEON)
+static inline
+unsigned char *(*rans_enc_func(int do_simd, int order))
+    (unsigned char *in,
+     unsigned int in_size,
+     unsigned char *out,
+     unsigned int *out_size) {
+
+    if (do_simd) {
+	if (rans_cpu & RANS_CPU_ENC_NEON)
+	    return order & 1
+		? rans_compress_O1_32x16_neon
+		: rans_compress_O0_32x16_neon;
+	else
+	    return order & 1
+		? rans_compress_O1_32x16
+		: rans_compress_O0_32x16;
+    } else {
+	return order & 1
+	    ? rans_compress_O1_4x16
+	    : rans_compress_O0_4x16;
+    }
+}
+
+static inline
+unsigned char *(*rans_dec_func(int do_simd, int order))
+    (unsigned char *in,
+     unsigned int in_size,
+     unsigned char *out,
+     unsigned int out_size) {
+
+    if (do_simd) {
+	if (rans_cpu & RANS_CPU_DEC_NEON)
+	    return order & 1
+		? rans_uncompress_O1_32x16_neon
+		: rans_uncompress_O0_32x16_neon;
+	else
+	    return order & 1
+		? rans_uncompress_O1_32x16
+		: rans_uncompress_O0_32x16;
+    } else {
+	return order & 1
+	    ? rans_uncompress_O1_4x16
+	    : rans_uncompress_O0_4x16;
+    }
+}
+
+#else // defined(__GNUC__) && defined(__x86_64__)
+
+// We may well be able to write generate AVX2 code, but if we can't auto-detect
+// it then it's pointless.  We can however still use the 32-way codec as
+// we may be decoding on a machine with AVX2 support and the CPU hit isn't
+// vast.
+#ifdef HAVE_AVX2
+#  undef HAVE_AVX2
+#endif
+
+#ifdef HAVE_AVX2
```

Comment:
I don't really understand what undefining these here achieves, but this one is surely a typo for `#ifdef HAVE_AVX512`.

Reply:
There were a few typos from code kill-and-yank operations of this nature, but I thought we'd corrected them all.

The undef seem to be a relic though so the entire thing could be culled.  When I initially started work on these I just had one source file that `#include`d the various others based on the existance of the definitions, so undefining them made sence.  Once I got further into the project I did it properly via Makefile rules and file-specific compiler opts (the only reliable way to work infact), but it looks like some left-overs weren't culled.